### PR TITLE
Latex bugs

### LINKED ISSUE #18669: DataFrame.to_latex() outputs empty cells if MultiIndex names are empty strings
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([[1, 2], [3,4]])

In [3]: df = pd.DataFrame(-1, index=mi, columns=range(4))

In [4]: df.index.names = ['', '']

In [5]: print(df.to_latex())
\begin{tabular}{llrrrr}
\toprule
  &   &  0 &  1 &  2 &  3 \\
\midrule
1 & 3 &    &    &    &    \\
  & 4 & -1 & -1 & -1 & -1 \\
2 & 3 & -1 & -1 & -1 & -1 \\
  & 4 & -1 & -1 & -1 & -1 \\
\bottomrule
\end{tabular}
```

#### Problem description

``df`` contains 4 rows; however the output of ``df.to_latex()`` only contains 3 rows of -1.

#### Expected Output

```python
\begin{tabular}{llrrrr}
\toprule
  &   &  0 &  1 &  2 &  3 \\
\midrule
1 & 3 & -1 & -1 & -1 & -1 \\
  & 4 & -1 & -1 & -1 & -1 \\
2 & 3 & -1 & -1 & -1 & -1 \\
  & 4 & -1 & -1 & -1 & -1 \\
\bottomrule
\end{tabular}
```

#### Output of ``pd.show_versions()``

<details>

In [7]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: fdba133336c2ec8f1269eb15e73d48797840c48b
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.22.0.dev0+301.gfdba13333
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.7.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### LINKED ISSUE #19981: to_latex() errors if index level name is not string
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.DataFrame([[1, 2, 3]]*2).set_index([0, 1]).to_latex()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-22b8b94dc46c> in <module>()
----> 1 pd.DataFrame([[1, 2, 3]]*2).set_index([0, 1]).to_latex()

/home/nobackup/repo/pandas/pandas/core/generic.py in to_latex(self, buf, columns, col_space, header, index, na_rep, formatters, float_format, sparsify, index_names, bold_rows, column_format, longtable, escape, encoding, decimal, multicolumn, multicolumn_format, multirow)
   2152                            encoding=encoding, multicolumn=multicolumn,
   2153                            multicolumn_format=multicolumn_format,
-> 2154                            multirow=multirow)
   2155 
   2156         if buf is None:

/home/nobackup/repo/pandas/pandas/io/formats/format.py in to_latex(self, column_format, longtable, encoding, multicolumn, multicolumn_format, multirow)
    709 
    710         if hasattr(self.buf, 'write'):
--> 711             latex_renderer.write_result(self.buf)
    712         elif isinstance(self.buf, compat.string_types):
    713             import codecs

/home/nobackup/repo/pandas/pandas/io/formats/format.py in write_result(self, buf)
    984                          .replace('}', '\\}').replace('~', '\\textasciitilde')
    985                          .replace('^', '\\textasciicircum').replace('&', '\\&')
--> 986                          if (x and x != '{}') else '{}') for x in row]
    987             else:
    988                 crow = [x if x else '{}' for x in row]

/home/nobackup/repo/pandas/pandas/io/formats/format.py in <listcomp>(.0)
    984                          .replace('}', '\\}').replace('~', '\\textasciitilde')
    985                          .replace('^', '\\textasciicircum').replace('&', '\\&')
--> 986                          if (x and x != '{}') else '{}') for x in row]
    987             else:
    988                 crow = [x if x else '{}' for x in row]

AttributeError: 'int' object has no attribute 'replace'

```

#### Problem description

``to_latex()`` assumes that ``MultiIndex`` level names, if any, are strings.

Related to #18669

As a general comment, I think ``to_latex()`` would benefit a lot from some refactoring which abstracted some of the (non-)escaping subtleties. There should be a way to avoid a ``replace()`` for each specific command, setting placeholders instead which are finally transformed all at once. This would also allow users to also insert "non-escapand" content (e.g. LaTeX formulae as labels) without setting ``escape=False``, which in many cases is an imperfect workaround.

#### Expected Output

```python
In [2]: pd.DataFrame([[1, 2, 3]]*2, columns=list('012')).set_index(['0', '1']).to_latex()
Out[2]: '\\begin{tabular}{llr}\n\\toprule\n  &   &  2 \\\\\n0 & 1 &    \\\\\n\\midrule\n1 & 2 &  3 \\\\\n  &   &  3 \\\\\n\\bottomrule\n\\end{tabular}\n'
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 0bfb61b2114769a29972ef90098c6dda005dbd23
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-5-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.23.0.dev0+422.g0bfb61b21.dirty
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.7.0
Cython: 0.25.2
numpy: 1.14.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by shangyian**
- [x] closes #19981 and 
closes #18669 
- [x] tests added / passed (see `test_to_latex.py`)
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -945,19 +927,22 @@ def get_col_type(dtype):
...
 
-        if not self.longtable:
-            buf.write('\\begin{{tabular}}{{{fmt}}}\n'
-                      .format(fmt=column_format))
-            buf.write('\\toprule\n')
-        else:
-            buf.write('\\begin{{longtable}}{{{fmt}}}\n'
-                      .format(fmt=column_format))
-            buf.write('\\toprule\n')
+    def write_result(self, buf):
```
Comment (shangyian) [inline pandas/io/formats/format.py]: I did some refactoring here and broke this method down into a couple of smaller ones. Ideally, pieces like `_rebuild_multi_index()` could also be a bit clearer, but maybe I'll tackle that while looking at some of the other Latex-related bugs.

**Comment by shangyian** [comment]
Also, after going through some of the other Latex-related bugs, I found [this discussion](https://github.com/pandas-dev/pandas/issues/13203#issuecomment-219744505) about refactoring `to_latex` to use `Styler`. @TomAugspurger, do you think that still makes sense?

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -975,38 +960,51 @@ def get_col_type(dtype):
...
+            # sum up columns to multicolumns
+            crow = self._format_multicolumn(crow, ilevels)
+        if (i >= nlevels and self.fmt.index and self.multirow and
+                ilevels > 1):
+            # sum up rows to multirows
+            crow = self._format_multirow(crow, ilevels, i, strrows)
+        return crow
+
+    def _escape_row(self, row):
+        def null_replace(x):
```
Comment (jreback) [inline pandas/io/formats/format.py]: can you add doc-strings to added methods

**Review by jreback — CHANGES_REQUESTED** [review]
@shangyian so if going to refactor (great idea!). Would you mind doing a pre-cursor PR to split out to separate files, csv.py, html.py, latex.py (just the CSVFormatter and like classes). into separate module. Then should be much cleaner to re-factor this (e.g. you can make module level & static functions easily)

**Comment by tomneep** [comment]
Hi @shangyian, 
I came across this because I've also been looking at some to_latex bugs and have a PR open: https://github.com/pandas-dev/pandas/pull/19910

As part of the discussion I also realised that what you've renamed ``_rebuild_multi_index()`` could be much cleaner. I had a go at this here https://github.com/tomneep/pandas/blob/to_latex_nan_fix_alternative/pandas/io/formats/format.py#L905 and think it really helps. I think it also addresses #19981  (as well as fixing a couple of other bugs). So maybe that is a potential option. 

I have a question about one of your new tests, ``test_to_latex_multiindex_non_string``, I'd expect the header to be `0 & 1 &    \\` but in your test you have `{} & 1 &    \\`. Am I missing something?

**Comment by shangyian** [comment]
@jreback - I did a preliminary PR to move the different formatters to separate files here: https://github.com/pandas-dev/pandas/pull/20051. Let me know what you think.

**Comment by shangyian** [comment]
@tomneep - Good catch on the test - not sure why I thought that made sense! As for the multi-indexing portion of the method, yeah, it makes sense to rework that to be clearer. I didn't change much from the original code when I pulled it into `_rebuild_multi_index`.

**Comment by jreback** [comment]
ok let's rebase this and see where we are

**Comment by shangyian** [comment]
Rebased and added some comments.

**Comment by shangyian** [comment]
Not sure if best to do this here, but we discussed combining `get_level_lengths` and `_get_level_lengths` in `formats.style` and I just did that now.

**Comment by jreback** [comment]
can you rebase / fixup

**Comment by jreback** [comment]
closing as stale. if you'd like to continue pls ping.

**Pull request closed without merging**